parser-typechecker/src/Unison/Codebase/SqliteCodebase.hs

```diff
@@ -1058,67 +1062,74 @@ viewRemoteBranch' (repo, sbh, path) action = UnliftIO.try do
-          (fmap . fmap) Cv.branchHash2to1 (runDB destConn Ops.loadMaybeRootCausalHash) >>= \case
-            Nothing -> do
-              setRepoRoot newRootHash
-              Q.release "push"
-              pure $ Right ()
-            Just oldRootHash -> do
-              before oldRootHash newRootHash >>= \case
-                Nothing ->
-                  error $
-                    "I couldn't find the hash " ++ show newRootHash
-                      ++ " that I just synced to the cached copy of "
-                      ++ repoString
-                      ++ " in "
-                      ++ show remotePath
-                      ++ "."
-                Just False -> do
-                  Q.rollbackRelease "push"
-                  pure . Left . C.GitProtocolError $ GitError.PushDestinationHasNewStuff repo
-                Just True -> do
-                  setRepoRoot newRootHash
-                  Q.release "push"
-                  pure $ Right ()
-        else do
-          Q.release "push"
-          pure $ Right ()
-
-      Q.setJournalMode JournalMode.DELETE
-      pure result
-  liftIO do
-    void $ push remotePath repo
+  pathToCachedRemote <- time "Git fetch" $ throwExceptTWith C.GitProtocolError $ pullRepo (writeToRead repo)
+  throwEitherMWith C.GitProtocolError . withIsolatedRepo pathToCachedRemote $ \isolatedRemotePath -> do
+    -- Connect to codebase in the cached git repo so we can copy it over.
+    withOpenOrCreateCodebaseConnection @m "push.cached" pathToCachedRemote $ \cachedRemoteConn -> do
+      -- Copy our cached remote database cleanly into our isolated directory.
+      copyCodebase cachedRemoteConn isolatedRemotePath
+      -- Connect to the newly copied database which we know has been properly closed and
+      -- nobody else could be using.
+      withConnection @m "push.dest" isolatedRemotePath $ \destConn -> do
+        flip runReaderT destConn $ Q.withSavepoint_ @(ReaderT _ m) "push" $ do
+          throwExceptT $ doSync isolatedRemotePath srcConn destConn
+    void $ push isolatedRemotePath repo
   where
+    doSync :: FilePath -> Connection -> Connection -> ExceptT C.GitError (ReaderT Connection m) ()
+    doSync remotePath srcConn destConn = do
+      _ <- flip State.execStateT emptySyncProgressState $
+        syncInternal syncProgress srcConn destConn (Branch.transform (lift . lift . lift) branch)
+      when setRoot $ overwriteRoot remotePath destConn
+    overwriteRoot :: forall m. MonadIO m => FilePath -> Connection -> ExceptT C.GitError m ()
+    overwriteRoot remotePath destConn = do
+      let newRootHash = Branch.headHash branch
+      -- the call to runDB "handles" the possible DB error by bombing
+      maybeOldRootHash <- fmap Cv.branchHash2to1 <$> runDB destConn Ops.loadMaybeRootCausalHash
+      case maybeOldRootHash of
+        Nothing -> runDB destConn $ do
+          setRepoRoot newRootHash
+          -- Q.release "push"
```

Comment:
Did you mean to keep the release/rollback comments?

Reply:
Nope; thanks! 😄